openvdb_cmd/vdb_tool/include/Tool.h

```diff
@@ -0,0 +1,2846 @@
+
+////////////////////////////////////////////////////////////////////////////////
+///
+/// @author Ken Museth
+///
+/// @file Tool.h
+///
+/// @brief This tool can combine any sequence of the of high-level tools available in openvdb.
+///        For instance, it can convert a sequence of polygon meshes and particles to level sets,
+///        and perform a large number of operations on these level set surfaces. It can also
+///        generate adaptive polygon meshes from level sets, render images and write particles,
+///        meshes or VDBs them to disk.
+///
+////////////////////////////////////////////////////////////////////////////////
+
+#ifndef VDB_TOOL_HAS_BEEN_INCLUDED
+#define VDB_TOOL_HAS_BEEN_INCLUDED
+
+#include <openvdb/openvdb.h>
+#include <openvdb/io/Stream.h>
+#include <openvdb/util/CpuTimer.h>
+#include <openvdb/util/Formats.h>
+#include <openvdb/tools/Composite.h>
+#include <openvdb/tools/FastSweeping.h>
+#include <openvdb/tools/LevelSetAdvect.h>
+#include <openvdb/tools/LevelSetSphere.h>
+#include <openvdb/tools/LevelSetFilter.h>
+#include <openvdb/tools/LevelSetPlatonic.h>
+#include <openvdb/tools/LevelSetRebuild.h>
+#include <openvdb/tools/LevelSetUtil.h>
+#include <openvdb/tools/RayIntersector.h>
+#include <openvdb/tools/RayTracer.h>
+#include <openvdb/tools/MeshToVolume.h>
+#include <openvdb/tools/ParticlesToLevelSet.h>
+#include <openvdb/tools/PointScatter.h>
+#include <openvdb/tools/PointsToMask.h>
+#include <openvdb/tools/Composite.h>
+#include <openvdb/tools/VolumeToMesh.h>
+#include <openvdb/tools/GridOperators.h>
+#include <openvdb/tools/GridTransformer.h>
+#include <openvdb/tools/FastSweeping.h>
+#include <openvdb/tools/Prune.h>
+#include <openvdb/tools/Clip.h>
+#include <openvdb/tools/Mask.h> // for tools::interiorMask()
+#include <openvdb/tools/MultiResGrid.h>
+#include <openvdb/tools/SignedFloodFill.h>
+#include <openvdb/points/PointConversion.h>
+#include <openvdb/points/PointCount.h>
+
+#ifdef VDB_TOOL_USE_NANO
+#include <nanovdb/NanoVDB.h>
+#include <nanovdb/util/IO.h>
+#include <nanovdb/util/OpenToNanoVDB.h>
+#include <nanovdb/util/NanoToOpenVDB.h>
+#endif
+
+#ifdef VDB_TOOL_USE_EXR
+#include <OpenEXR/ImfChannelList.h>
+#include <OpenEXR/ImfFrameBuffer.h>
+#include <OpenEXR/ImfHeader.h>
+#include <OpenEXR/ImfOutputFile.h>
+#include <OpenEXR/ImfPixelType.h>
+#endif
+
+#ifdef VDB_TOOL_USE_PNG
+#include <png.h>
+#endif
+
+#ifdef VDB_TOOL_USE_JPG
+#include <jpeglib.h>
+#endif
+
+#include "Parser.h"
+#include "Geometry.h"
+
+#if defined(_WIN32)
+#include <io.h>
+#else
+#include <unistd.h>
+#endif
+
+namespace openvdb {
+OPENVDB_USE_VERSION_NAMESPACE
+namespace OPENVDB_VERSION_NAME {
+namespace vdb_tool {
+
+class Tool
+{
+    static const int sMajor =10;// incremented for incompatible changes options or file.
+    static const int sMinor = 6;// incremented for new functionality that is backwards-compatible.
+    static const int sPatch = 1;// incremented for backwards-compatible bug fixes.
+
```

Comment:
`sMinor` probably should mirror the OpenVDB's minor version, which is 0. `sPatch` also should be 0.

Reply:
unfortunately the version number for this tool has nothing to do with the version number of openvdb itself. Recall, OpenVDB is using semantic version numbering based on the Grid only and not any of the tools. That is to say the major number changes when the ABI of openvdb::Grid (or many of its dependent classes) change. However, the version number of vdb_tool has to do with backwards comparability of the the actions and option that it expose. Another what to say this is if two configuration files of vdb_tool share the same major version number then they are guaranteed to produce the same outcome and more importantly, if the major version of a binary vdb_tool does not match the major version of a configuration file then it will throw an error.